File: go/pondsocket/middleware.go
package pondsocket

import (
	"context"
	"sync"
)

type middleware[Request any, Response any] struct {
	handlers []handlerFunc[Request, Response]
	mutex    sync.RWMutex
}

func newMiddleWare[Request any, Response any]() *middleware[Request, Response] {
	return &middleware[Request, Response]{
		handlers: make([]handlerFunc[Request, Response], 0),
	}
}

func (m *middleware[Request, Response]) Use(handlers ...handlerFunc[Request, Response]) {
	m.mutex.Lock()

	defer m.mutex.Unlock()

	m.handlers = append(m.handlers, handlers...)
}

func (m *middleware[Request, Response]) Compose(others ...*middleware[Request, Response]) *middleware[Request, Response] {
	result := newMiddleWare[Request, Response]()

	m.mutex.RLock()

	handlersCopy := make([]handlerFunc[Request, Response], len(m.handlers))

	copy(handlersCopy, m.handlers)

	m.mutex.RUnlock()

	result.Use(handlersCopy...)

	for _, other := range others {
		other.mutex.RLock()

		otherHandlersCopy := make([]handlerFunc[Request, Response], len(other.handlers))

		copy(otherHandlersCopy, other.handlers)

		other.mutex.RUnlock()

		result.Use(otherHandlersCopy...)
	}
	return result
}

func executeWithMiddleware[C any](ctx *C, handler func(ctx *C) error, middlewares []MiddlewareFunc[C]) error {
	if len(middlewares) == 0 {
		return handler(ctx)
	}
	var run func(i int) error
	run = func(i int) error {
		if i >= len(middlewares) {
			return handler(ctx)
		}
		return middlewares[i](ctx, func() error {
			return run(i + 1)
		})
	}
	return run(0)
}

func (m *middleware[Request, Response]) Handle(ctx context.Context, request Request, response Response, finalHandler FinalHandlerFunc[Request, Response]) error {
	select {
	case <-ctx.Done():
		return ctx.Err()

	default:
	}
	m.mutex.RLock()

	handlersCopy := make([]handlerFunc[Request, Response], len(m.handlers))

	copy(handlersCopy, m.handlers)

	m.mutex.RUnlock()

	if len(handlersCopy) == 0 {
		return finalHandler(request, response)
	}

	var executeHandler func(index int) error
	executeHandler = func(index int) error {
		select {
		case <-ctx.Done():
			return ctx.Err()

		default:
		}
		if index >= len(handlersCopy) {
			return finalHandler(request, response)
		}
		handler := handlersCopy[index]
		next := func() error {
			return executeHandler(index + 1)
		}
		return handler(ctx, request, response, next)
	}
	return executeHandler(0)
}

File: python/pondsocket/src/pondsocket/middleware.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from typing import Generic, TypeAlias, TypeVar

from .errors import PondError, wrap

Request = TypeVar("Request")
Response = TypeVar("Response")

NextFn: TypeAlias = Callable[[], Awaitable[None]]
HandlerFunc: TypeAlias = Callable[[Request, Response, NextFn], Awaitable[None]]
FinalHandlerFunc: TypeAlias = Callable[[Request, Response], Awaitable[None]]


class Middleware(Generic[Request, Response]):
    __slots__ = ("_handlers", "_lock")

    def __init__(self) -> None:
        self._handlers: list[HandlerFunc[Request, Response]] = []
        self._lock = asyncio.Lock()

    async def use(self, *handlers: HandlerFunc[Request, Response]) -> None:
        async with self._lock:
            self._handlers.extend(handlers)

    def use_sync(self, *handlers: HandlerFunc[Request, Response]) -> None:
        self._handlers.extend(handlers)

    async def compose(
        self, *others: Middleware[Request, Response]
    ) -> Middleware[Request, Response]:
        result: Middleware[Request, Response] = Middleware()
        async with self._lock:
            result._handlers.extend(self._handlers)
        for other in others:
            async with other._lock:
                result._handlers.extend(other._handlers)
        return result

    async def handle(
        self,
        request: Request,
        response: Response,
        final: FinalHandlerFunc[Request, Response],
    ) -> None:
        async with self._lock:
            handlers = list(self._handlers)

        if not handlers:
            await final(request, response)
            return

        async def run(index: int) -> None:
            if index >= len(handlers):
                await final(request, response)
                return

            async def next_fn() -> None:
                await run(index + 1)

            await handlers[index](request, response, next_fn)

        await run(0)

    def __len__(self) -> int:
        return len(self._handlers)


C = TypeVar("C")
PerHandlerMiddleware: TypeAlias = Callable[[C, NextFn], Awaitable[None]]


async def execute_with_middleware(
    ctx: C,
    handler: Callable[[C], Awaitable[None]],
    middlewares: list[PerHandlerMiddleware[C]],
) -> None:
    if not middlewares:
        await handler(ctx)
        return

    async def run(index: int) -> None:
        if index >= len(middlewares):
            await handler(ctx)
            return

        async def next_fn() -> None:
            await run(index + 1)

        await middlewares[index](ctx, next_fn)

    await run(0)


def wrap_pond_error(err: BaseException, message: str) -> PondError:
    if isinstance(err, PondError):
        return wrap(err, message)
    return wrap(err, message)
